### PR TITLE
Metadata based healthcheck

### PR DESCRIPTION
**Pull request by wlan0**
@ibuildthecloud

### CODE REVIEW

File: storagepool/fileBasedHealthCheck.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+	log "github.com/Sirupsen/logrus"
+)
+
+type fileBasedHealthCheck struct {
+	healthCheckBaseDir string
+}
+
+func (fc *fileBasedHealthCheck) populateHostMap() (map[string]string, error) {
+	activeHosts := map[string]string{}
+	f, err := os.Open(fc.healthCheckBaseDir)
```
Comment (ibuildthecloud) [inline storagepool/fileBasedHealthCheck.go]: you need to defer close here

File: storagepool/fileBasedHealthCheck.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+	if err != nil {
+		return nil, err
+	}
+	fi, err := f.Readdir(-1)
```
Comment (ibuildthecloud) [inline storagepool/fileBasedHealthCheck.go]: use https://golang.org/pkg/io/ioutil/#ReadDir

File: storagepool/agent.go [hunk 3]
```diff
@@ -17,20 +17,22 @@ type StoragepoolAgent struct {
 	storagepoolRootDir  string
 	storagepoolUuid     string
 	healthCheckBaseDir  string
+	healthCheckType     string
 	cattleClient        cattle.CattleInterface
 }
 
-func NewStoragepoolAgent(healthCheckInterval int, storagepoolRootDir, storagepoolUuid, healthCheckBaseDir string, cattleClient cattle.CattleInterface) *StoragepoolAgent {
+func NewStoragepoolAgent(healthCheckInterval int, storagepoolRootDir, storagepoolUuid, healthCheckBaseDir, healthCheckType string, cattleClient cattle.CattleInterface) *StoragepoolAgent {
```
Comment (ibuildthecloud) [inline storagepool/agent.go]: Just pass in the health checker object, not the type.  the caller can construct the health checker.

File: storagepool/fileBasedHealthCheck.go (hunk 2 above)
Comment (wlan0) [inline storagepool/fileBasedHealthCheck.go]: It does the same thing that im doing -   

```
   99 // a list of sorted directory entries.
   100  func ReadDir(dirname string) ([]os.FileInfo, error) {
   101      f, err := os.Open(dirname)
   102      if err != nil {
   103          return nil, err
   104      }
   105      list, err := f.Readdir(-1)
   106      f.Close()
   107      if err != nil {
   108          return nil, err
   109      }
   110      sort.Sort(byName(list))
   111      return list, nil
   112  }
   113  
```

I'll just add f.Close()?

File: storagepool/metadataBasedHealthCheck.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+	services, err := m.GetServices()
+	if err != nil {
+		return nil, err
+	}
+
+	possibleContainers := []string{}
+
+	for _, service := range services {
+		if contains(serviceNames, service.Name) && service.StackName == stackName {
```
Comment (ibuildthecloud) [inline storagepool/metadataBasedHealthCheck.go]: `contains(serviceNames, service.Name)` is redundant, `service.StackName == stackName` is sufficient

File: storagepool/metadataBasedHealthCheck.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+			possibleContainers = append(possibleContainers, service.Containers...)
+		}
+	}
+
+	containers, err := m.GetContainers()
+	if err != nil {
+		return nil, err
+	}
+	for _, container := range containers {
```
Comment (ibuildthecloud) [inline storagepool/metadataBasedHealthCheck.go:61]: this O(n^2)

File: storagepool/metadataBasedHealthCheck.go [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+		if contains(possibleContainers, container.Name) && container.StackName == stackName && contains(serviceNames, container.ServiceName) {
```
Comment (ibuildthecloud) [inline storagepool/metadataBasedHealthCheck.go]: `contains(serviceNames, container.ServiceName)` is not needed

File: storagepool/storagepool.go [hunk 7]
```diff
@@ -49,9 +55,11 @@ func storagepoolAgent(c *cli.Context) {
 		log.Fatal(err)
 	}
 
-	storagepoolAgent := NewStoragepoolAgent(healthCheckInterval, storagepoolRootDir, storagepoolUUID, healthCheckBaseDir, cattleClient)
+	storagepoolAgent := NewStoragepoolAgent(healthCheckInterval, storagepoolRootDir, storagepoolUUID, healthCheckBaseDir, healthCheckType, cattleClient)
+
+	metadataUrl := "http://rancher-metadata"
```
Comment (ibuildthecloud) [inline storagepool/storagepool.go]: Take this as an option to the program and default it to `http://rancher-metadata/07-25-2015`

File: tests/metadata_server.go [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+	selfStack  = metadata.Stack{}
+	services   = []metadata.Service{}
+	containers = []metadata.Container{}
+
+	metadataUrl = ":12345"
+)
+
+func startMetadataServer() {
+	mux := http.NewServeMux()
+	mux.HandleFunc("/latest/version", func(w http.ResponseWriter, req *http.Request) {
```
Comment (ibuildthecloud) [inline tests/metadata_server.go]: Switch to the versioned URL `07-25-2015`

File: storagepool/metadataBasedHealthCheck.go (hunk 5 above)
Comment (wlan0) [inline storagepool/metadataBasedHealthCheck.go:61]: These are pseudo linear now, im using a map.

File: storagepool/fileBasedHealthCheck.go (hunk 2 above)
Comment (wlan0) [inline storagepool/fileBasedHealthCheck.go]: added f.Close()

Comment (ibuildthecloud) [inline storagepool/fileBasedHealthCheck.go]: Great minds think a like.  Use ReadDir

**Pull request merged**